### PR TITLE
Synchronize brave vpn system tray with the browser and the wireguard service

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/30708

- Added main menu item to show/hide tray icon for brave vpn
- Added menu to the tray icon to manage wireguard vpn connection
- Updated the service watcher to be able to subscribe to different events by mask
- Auto start tray icon process when wireguard vpn connected
   - The tray process manages cases to show/hide icon inside itself
   - It checks if already started or user made icon hidden or no last used config yet.

https://github.com/brave/brave-core/assets/2965009/3066386c-9c22-45a8-8b4b-c946ca72a5a3

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Enable wireguard vpn brave://flags/#brave-vpn-wireguard
- Connect to BraveVpn in the browser and check the tray place for the BraveVpn icon.
- Play with the tray icon menu and with the browser vpn connection toggler, try to connect/disconnect. The tray icon should show all actual statuses for the connection.
- Check for different windows themes, the dark theme should have own icons.
- If user clicked to remove the tray icon then it can be enabled back from the browser main submenu for BraveVpn.

### CODE REVIEW

File: components/brave_vpn/common/wireguard/win/storage_utils.cc [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+          HKEY_CURRENT_USER,
+          brave_vpn::GetBraveVpnWireguardServiceRegistryStoragePath().c_str(),
+          KEY_QUERY_VALUE) != ERROR_SUCCESS) {
+    return true;
+  }
+  DWORD value = 1;
+  if (storage.ReadValueDW(kBraveWireguardEnableTrayIconKeyName, &value) !=
+      ERROR_SUCCESS) {
+    return true;
+  }
```
Comment (simonhong) [inline components/brave_vpn/common/wireguard/win/storage_utils.cc:33]: We return `true` when reg operation failed. Is this intended?

File: browser/ui/brave_browser_command_controller.cc [hunk 2]
```diff
@@ -249,15 +249,16 @@ void BraveBrowserCommandController::UpdateCommandForBraveVPN() {
...
     UpdateCommandEnabled(IDC_SEND_BRAVE_VPN_FEEDBACK, false);
     UpdateCommandEnabled(IDC_ABOUT_BRAVE_VPN, false);
     UpdateCommandEnabled(IDC_MANAGE_BRAVE_VPN_PLAN, false);
     UpdateCommandEnabled(IDC_TOGGLE_BRAVE_VPN, false);
     return;
   }
-
   UpdateCommandEnabled(IDC_SHOW_BRAVE_VPN_PANEL, true);
   UpdateCommandEnabled(IDC_TOGGLE_BRAVE_VPN_TOOLBAR_BUTTON, true);
+  UpdateCommandEnabled(IDC_TOGGLE_BRAVE_VPN_TRAY_ICON, true);
```
Comment (simonhong) [inline browser/ui/brave_browser_command_controller.cc:261]: This command should be enabled only when wireguard feature is enabled?

File: browser/ui/toolbar/brave_vpn_menu_model.cc [hunk 3]
```diff
@@ -29,6 +34,12 @@ void BraveVPNMenuModel::Build() {
                       IsBraveVPNButtonVisible()
                           ? IDS_BRAVE_VPN_HIDE_VPN_BUTTON_MENU_ITEM
                           : IDS_BRAVE_VPN_SHOW_VPN_BUTTON_MENU_ITEM);
+#if BUILDFLAG(IS_WIN)
+  AddItemWithStringId(IDC_TOGGLE_BRAVE_VPN_TRAY_ICON,
```
Comment (simonhong) [inline browser/ui/toolbar/brave_vpn_menu_model.cc:38]: same. I think this entry also should be shown only when wg feature is enabled?

File: browser/ui/toolbar/brave_vpn_menu_model.h [hunk 4]
```diff
@@ -7,26 +7,39 @@
...
 #include "ui/base/models/simple_menu_model.h"
 
 class Browser;
+class PrefService;
 
 class BraveVPNMenuModel : public ui::SimpleMenuModel,
                           public ui::SimpleMenuModel::Delegate {
  public:
-  explicit BraveVPNMenuModel(Browser* browser);
+  explicit BraveVPNMenuModel(Browser* browser, PrefService* profile_prefs);
```
Comment (simonhong) [inline browser/ui/toolbar/brave_vpn_menu_model.h:19]:
```suggestion
  BraveVPNMenuModel(Browser* browser, PrefService* profile_prefs);
```

File: components/brave_vpn/common/wireguard/win/storage_utils.cc (hunk 1 above)
Comment (spylogsster) [reply components/brave_vpn/common/wireguard/win/storage_utils.cc:33]: yes, it is disabled only if we have zero set

File: browser/ui/toolbar/brave_vpn_menu_model.cc (hunk 3 above)
Comment (spylogsster) [reply browser/ui/toolbar/brave_vpn_menu_model.cc]: done

File: browser/ui/brave_browser_command_controller.cc (hunk 2 above)
Comment (spylogsster) [reply browser/ui/brave_browser_command_controller.cc]: done

**Review by simonhong — APPROVED** [review]
++ 👍🏼

File: app/brave_generated_resources.grd [hunk 5]
```diff
@@ -996,6 +996,12 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
           <message name="IDS_BRAVE_VPN_HIDE_VPN_BUTTON_MENU_ITEM" desc="The menu item for hiding vpn button in toolbar">
             Hide VPN Button in Toolbar
           </message>
+          <message name="IDS_BRAVE_VPN_SHOW_VPN_TRAY_ICON_MENU_ITEM" desc="The menu item for showing vpn tray icon in windows">
+            Show VPN Tray Icon
+          </message>
+          <message name="IDS_BRAVE_VPN_HIDE_VPN_TRAY_ICON_MENU_ITEM" desc="The menu item for hiding  vpn tray icon in windows">
```
Comment (mkarolin) [inline app/brave_generated_resources.grd:1002]: There's an extra space between `hiding` and `vpn`

File: app/brave_generated_resources.grd [hunk 6]
```diff
@@ -1010,6 +1016,12 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
           <message name="IDS_BRAVE_VPN_HIDE_VPN_BUTTON_MENU_ITEM" desc="The menu item for hiding vpn button in toolbar">
             Hide VPN button in toolbar
           </message>
+          <message name="IDS_BRAVE_VPN_SHOW_VPN_TRAY_ICON_MENU_ITEM" desc="The menu item for showing vpn tray icon in windows">
+            Show VPN tray icon
+          </message>
+          <message name="IDS_BRAVE_VPN_HIDE_VPN_TRAY_ICON_MENU_ITEM" desc="The menu item for hiding  vpn tray icon in windows">
```
Comment (mkarolin) [inline app/brave_generated_resources.grd:1022]: There's an extra space between `hiding` and `vpn`

**Review by mkarolin — APPROVED** [review]
`strings` ++ (with 2 formatting nits).

File: app/brave_generated_resources.grd (hunk 5 above)
Comment (spylogsster) [reply app/brave_generated_resources.grd]: done

File: app/brave_generated_resources.grd (hunk 6 above)
Comment (spylogsster) [reply app/brave_generated_resources.grd]: done

**Pull request merged**